### PR TITLE
Provide progress updates for running reindexes.

### PR DESCRIPTION
**Pull request by LATaylor-guardian**
This change implements reindex progress updates against a fake endpoint inside that
application for now that fakes the response that we would expect to receive from a potential
client.

[Polling actor modelled on @cb372's really nice implementation in Pubflow]

### CODE REVIEW

File: app/com/gu/floodgate/DynamoDBTable.scala [hunk 1]
```diff
@@ -111,15 +113,31 @@ trait DynamoDBTable[T] extends StrictLogging {
...
     val request = new UpdateItemRequest().withTableName(tableName)
-      .withKey(Map("id" -> new AttributeValue(id)))
+      .withKey(Map(keyName -> new AttributeValue(id)))
       .withAttributeUpdates(toItemUpdate(t).mapValues(_.withAction("PUT")))
 
     dynamoDB.updateItemAsync(request)
   }
 
-  def deleteItem(id: String) = {
+  /* TODO - implement correctly with return type. */
```
Comment (cb372) [inline app/com/gu/floodgate/DynamoDBTable.scala]: It's already returning a `Future[DeleteItemResult]` so you could just change the return type to that? Seems like a reasonable thing to return.

File: app/com/gu/floodgate/reindex/Progress.scala [hunk 2]
```diff
@@ -0,0 +1,6 @@
+package com.gu.floodgate.reindex
+
+import play.json.extra.JsonFormat
+
+@JsonFormat
+case class Progress(status: String, documentsIndexed: Int, documentsExpected: Int)
```
Comment (cb372) [inline app/com/gu/floodgate/reindex/Progress.scala]: Could we make the status an enum/sealed trait? Or do you want to support arbitrary status messages? If you want to keep it as a string, could we at least make it something that looks a little more machine-readable than `"in progress"`?

File: app/com/gu/floodgate/reindex/ProgressTracker.scala [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+          case _ => onFailure(contentSource, runningJob)
+        }
+      case Failure(e) => onFailure(contentSource, runningJob)
+    }
+  }
+
+  private def onSuccess(response: WSResponse, contentSource: ContentSource, runningJob: RunningJob): Unit = {
+    response.json.validate[Progress].fold(
+      error => {
+        logger.warn(s"Content source with id: ${contentSource.id} appears to be returning progress updates in an incorrect format. Cancelling reindex.")
```
Comment (cb372) [inline app/com/gu/floodgate/reindex/ProgressTracker.scala]: Not strictly correct. We don't cancel the reindex, we just mark it as cancelled and stop monitoring it.

File: app/com/gu/floodgate/reindex/ProgressTrackerController.scala [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+object ProgressTrackerController {
+  def props(ws: WSAPI, runningJobService: RunningJobService, jobHistoryService: JobHistoryService) = Props(new ProgressTrackerController(ws: WSAPI, runningJobService: RunningJobService, jobHistoryService))
+
+  case class LaunchTracker(contentSource: ContentSource, runningJob: RunningJob)
+  case class RemoveTracker(contentSource: ContentSource, runningJob: RunningJob)
+}
+
+class ProgressTrackerController(ws: WSAPI, runningJobService: RunningJobService, jobHistoryService: JobHistoryService) extends Actor with ActorLogging with StrictLogging {
+
+  private var runningTrackers: Map[String, ActorRef] = Map.empty[String, ActorRef]
```
Comment (cb372) [inline app/com/gu/floodgate/reindex/ProgressTrackerController.scala]: Is there any reason to keep tracking of the running actors?

File: app/com/gu/floodgate/runningjob/RunningJobTable.scala [hunk 5]
```diff
@@ -19,17 +20,19 @@ class RunningJobTable(protected val dynamoDB: AmazonDynamoDBAsyncClient, protect
   override protected def fromItem(item: Map[String, AttributeValue]): RunningJob =
     RunningJob(
       getItemAttributeValue(fields.ContentSourceId, item).getS,
-      getItemAttributeValue(fields.Progress, item).getN.asInstanceOf[Int], // why does toInt not work?
-      new DateTime(getItemAttributeValue(fields.StartTime, item).getS)
-    )
+      getItemAttributeValue(fields.DocumentsIndexed, item).getS.toInt,
```
Comment (cb372) [inline app/com/gu/floodgate/runningjob/RunningJobTable.scala]: Storing an integer as an `S` is a shame. I think if you store it as an `N` then you might have to convert it to a `BigDecimal` and then an `Int`. That's what the DynamoDB Item API does.

File: app/com/gu/floodgate/reindex/ProgressTrackerController.scala (hunk 4 above)
Comment (LATaylor-guardian) [inline app/com/gu/floodgate/reindex/ProgressTrackerController.scala]: Yes, when we cancel a reindex via the API, we need to cancel the actor monitoring the progress.

**Comment by cb372** [comment]
:+1:

**Pull request merged**